File: leetcode/21.py
"""
Runtime: 32 ms, faster than 90.36% of Python3 online submissions for Merge Two Sorted Lists.
Memory Usage: 14.2 MB, less than 60.99% of Python3 online submissions for Merge Two Sorted Lists.
"""
# iteration
class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        head = ListNode(-1)
        dummy = head
        while l1 and l2:
            if l1.val < l2.val:
                head.next = l1
                l1 = l1.next
            else:
                head.next = l2
                l2 = l2.next
            head = head.next

        head.next = l1 if l1 else l2

        return dummy.next


"""
Runtime: 40 ms, faster than 43.05% of Python3 online submissions for Merge Two Sorted Lists.
Memory Usage: 14.4 MB, less than 29.56% of Python3 online submissions for Merge Two Sorted Lists.
"""
# recursion
class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummy = ListNode(-1)

        # exit case
        if not l1 or not l2:
            return l1 if l1 else l2

        # have two link list
        if l1.val < l2.val:
            dummy.next = l1
            l1.next = self.mergeTwoLists(l1.next, l2)
        else:
            dummy.next = l2
            l2.next = self.mergeTwoLists(l1, l2.next)

        return dummy.next


"""
Runtime: 40 ms, faster than 42.94% of Python3 online submissions for Merge Two Sorted Lists.
Memory Usage: 14 MB, less than 95.44% of Python3 online submissions for Merge Two Sorted Lists.
"""
# iteration using in-place
class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        if not l1 or not l2:
            return l1 or l2

        dummy = curr = ListNode(-1)
        curr.next = l1
        while l1 and l2:
            if l1.val < l2.val:
                l1 = l1.next
                curr = curr.next
            else:
                tmp1 = curr.next  # l1 tail
                curr.next = l2
                tmp2 = l2.next  # l2 tail
                curr = curr.next
                curr.next = tmp1
                l2 = tmp2

        curr.next = l1 or l2

        return dummy.next

File: leetcode/21.cpp
// Runtime: 7 ms
// Memory Usage: 15.1 MB
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
 public:
  ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
    ListNode* dummy = new ListNode(-1);
    ListNode* curr = dummy;
    while (list1 != nullptr && list2 != nullptr) {
      if (list1->val < list2->val) {
        curr->next = list1;
        list1 = list1->next;
      } else {
        curr->next = list2;
        list2 = list2->next;
      } 
      curr = curr->next;
    }
    curr->next = list1 != nullptr? list1 : list2;
    return dummy->next;
  }
};

// Runtime: 15 ms, faster than 35.22% of C++ online submissions for Merge Two Sorted Lists.
// Memory Usage: 14.9 MB, less than 42.51% of C++ online submissions for Merge Two Sorted Lists.
class Solution {
 public:
  ListNode *mergeTwoLists(ListNode *list1, ListNode *list2) {
    // given:
    //   - 2 linked lists, sorted
    //
    // goal:
    //   - combining the linked lists into single sorted
    //   - don't want to allocate extra memory
    //
    // approach:
    //   - track the tail of the sorted list
    //   - from list1, list2, see which value is closer and add it
    //   - increment the tail, and whichever list1 list2 was added

    // every value will be greater than dummy head
    ListNode *dummy = new ListNode(-101);
    ListNode *curr = dummy;

    while (list1 != nullptr && list2 != nullptr) {
      if ((list1->val - curr->val) < (list2->val - curr->val)) {
        curr->next = list1;
        list1 = list1->next;
        curr = curr->next;
      } else {
        curr->next = list2;
        list2 = list2->next;
        curr = curr->next;
      }
    }
    curr->next = list1 != nullptr ? list1 : list2;

    return dummy->next;
  }
};
